### DISCUSSION TITLE
Mackie Protocol - Current Channel Selected Mute/Solo/Record/Pan/Volume

### DISCUSSION THREAD

**Discussion by hali1850** [tttapa/Control-Surface#735] [Q&A]
Hi,

I have implemented the Mackie Control protocol (MCU::) in my code and everything is communicating well on my 1 fader, 1 potentiometer controller. 
However, I'm struggling to get around what I assume is either a limitation of the Mackie Control protocol, or a mistake on my part as I can't figure out how to do it with the resources of this Control-Surface library as yet.

As of right now, I'm using MCU::SOLO_1, MCU::MUTE_1, and etc to be able to use these functions in my daw. However, this all ofc only affects channel 1 in the session, rather than the channel which I have currently selected by mouse, or by my UP and DOWN channel selector buttons on this Teensy. 

In this video of a similar type of controller, which is also using MCU and is setup as an MCU Pro in the DAW, the controller is able to navigate the channels and apply mute and solo to the currently selected channel. 
https://www.youtube.com/watch?v=SOHi26sqOks @0:39
I can't see a function which can resemble this behaviour in the MCU section of this library.  

Is it possible to do this with this library? And if so, how so? 
And if not, could someone point me in the right direction possibly? It seems like such a simple thing... 
I don't want to believe that this is some custom firmware/proprietary/sysex/OSC stuff. At least I hope not.

Any help at all appreciated.
Many thanks

**Comment by Veroledez** [comment]
Hola, yo de nuevo.

Bien,si solo usas el primer fader y knob  o pan, no uses las direcciones MCU::BANK_, tienes que usar las MCU::CHANNEL_   , porque las de banco van de 8 en 8, por lo tanto pasaras de estar manrjando  la pista 1 a manejar la pista 9, mientras que con channel se desplazara  una pista a la vez.
Para que se seleccione automatico con el mouse,  cuando seleccionas una pista se envia algun mensaje, ese capturalo y y haz que se cambie los bancos o canal

**Reply by hali1850** [reply]
Hola Veroledez,

Thank you kindly for all of your help with my project, I really appreciate it.
Unfortunately I'm not quite sure what you mean by using MCU:CHANNEL_ instead of MCU::BANK. I have not used MCU:BANK. 
I am taking my information from this list of definitions: https://tttapa.github.io/Control-Surface-doc/Doxygen/db/d60/MCU_8hpp.html  

Here is my code: 
`#include <Control_Surface.h> // Include the Control Surface library
 
// Instantiate a MIDI over USB interface.
USBMIDI_Interface midi;
 
// Instantiate a CCButton object
NoteButton button1 {
  3,
  {MCU::PLAY, CHANNEL_1}, // address of first button
};
NoteButton button2  {
  6,
  {MCU::RECORD, CHANNEL_1}, // address of first button
};
NoteButton button3  {
  9,
  {MCU::STOP, CHANNEL_1}, // address of first button
};
NoteButton button4  {
  5,
  {MCU::CYCLE, CHANNEL_1}, // address of first button
};
NoteButton button5  {
  12,
  {MCU::MUTE_1, CHANNEL_1}, // address of first button
};
NoteButton button6  {
  2,
  {MCU::SOLO_1, CHANNEL_1}, // address of first button
}; 
NoteButton button7  {
  0,
  {MCU::UP, CHANNEL_1}, // address of first button
};
NoteButton button8  {
  1,
  {MCU::DOWN, CHANNEL_1}, // address of first button
}; 
NoteButton button9  {
  4,
  {MCU::AUTOMATION_LATCH, CHANNEL_1}, // address of first button
};
NoteButton button10  {
  7,
  {MCU::AUTOMATION_READ_OFF, CHANNEL_1}, // address of first button
};           

 
// Instantiate a PBPotentiometer object
PBPotentiometer potentiometer {
  A8,        // Analog pin connected to potentiometer
  CHANNEL_1, // MIDI Channel 1
};`

As you can see, I've placed in 'MUTE_1' because there is no definition in the list for a 'general' mute, of the currently selected track. Rather, the button is forever locked to only muting the first channel in the session, or at least the first channel in the currently selected bank of 8 that is. This is not the desired behaviour I'm after. 
You'll also see my PBPotentiometer (1 for now) which is supposed to be controlling the currently selected channel, instead is also locked to Channel 1. This is because I could also once again not find a 'general' volume definition/term for this- so for now it is just CHANNELl_1 unfortunately. So it too is forever locked to one specific channel, which again is not the desired behaviour. 

To clarify, I'm trying to achieve the ability for my mute, solo, pan and volume functions to affect the currently selected channel within the DAW. 

Another thing also, in the list of definitions there does seem to be a 'general' SOLO which does not have a channel number attached to it, however in my Arduino/Teensyduino terminal, 'SOLO' does not turn blue. Meaning it is not recognised as an actual definition. 
It is on line 129 at https://tttapa.github.io/Control-Surface-doc/Doxygen/db/d60/MCU_8hpp_source.html

I wonder if this has anything to do with what I'm after? 
Please further clarify what you meant specifically in your response my friend, so I can change my code. 

Also, you said 'For it to be selected automatically with the mouse, when you select a track a message is sent, capture it and make it change the banks or channel'. Once I know what message is sent from my DAW, how do I then 'make it change the banks or channel', in terms of code? Is there example code for this? I wouldn't know where to start with this in all honesty. I would greatly appreciate some insight on this also. 

Many thanks mi amigo. Hope you are well!

**Comment by Veroledez** [comment]
Hola, debes usar BANK_RIGHT BANKT_LEFT  CHANNEL_RIGHT.  CHANNEL_LEFT, los channel  desplazan de pista en pista estos usalos  en tu controlador, los bank desplazan de 8 en 8.

Para ayudarte con lo del daw, postea todos los mensajes que se envian desde el daw hacia teensy cuando seleccionas una pista, primero asegurate de que por ejemplo el Solo_1 afecte al track 1, luego con el mouse selecciona la pista 10, y captura los mensajes que se envian al seleccionar la pista 10.

Uso como daw  fl studio, y en este con mackie no hay mensajes cuando selecciono una pista que este por fuera de los 8, es decir si estoy controlando las primeras 8 pistas, si con el mouse selecciono de la 9 en adelante no hay ningun mensaje, pero si selecciono cualquiera dentro de las primeras 8 si hay varios mensajes, tienes que probar primero si hay mensajes

**Comment by rogerarends** [comment]
Hi
I used this table to implement the controls you are attempting to use.
Successfully created something simular to what you want about 2 years ago
and a 4 channel unit with screens. Although I haven't tried using it in
logic, only studio one v4 and and old version of Cubase. If I get a change
I'll try and find the code I used and email it to you.

**Comment by rogerarends** [comment]
Sorry couldn't find the example I wanted to show you, but have a look at
the first file (Mega) . I made it as basic as possible back then so I could
understand what I was doing and wanted to achieve. I'm not a programmer and
am not willing to learn at my age, anyway I tested it using a mega 2560 to
see if I could get all the functions to work properly and consistently then
on a teensy 4.
The second file was my first attempt to get 8 oled screens to work and it
did perfectly as a MCU controller. I used encoders instead of pots and used
the shift function to switch between volume and pan. The soft take over
worked perfectly as well with the encoders.

Coming back to your problem, I had to use a select button to change focus
to the track because the original Mackie had a touch sensitive fader. So
when you touch it, it automatically selects and focuses on that track. So I
think you should start with adding a select button first. If it doesn't
solve your problem then have a look at my mess of a code.
I'm going to search my old drives this weekend to see if I can find the
newer project files.

**Reply by hali1850** [reply]
Rogerarends,

Thank you ever so much for your contribution and offer of help. I'm appreciative you'd like to share your code also, that would be greatly helpful and I'm sure illuminating.
Unfortunately I did not receive an email from yourself as mentioned. My email is haiderali.1850@gmail.com.

As regards to the 'select' button, while it is a good workaround I'm afraid it's not as graceful as perhaps I'm after for my personal engineering workflow. 
Conversely, as the touch sensitive fader function is not required on a one fader, non-motorized design, I'm stuck to find a simple solution for replicating the desired behaviour for now. 
The quest continues..

Warm regards
